### PR TITLE
Parse classfile generic parameters defined in supertype generic signature

### LINKED ISSUE #26312: Classfile parser breaks on nested generics in dependency of community build
## Compiler version
3.8.3, also Scala 3.10.0-RC1-bin-20260609-b34a019-NIGHTLY

## Minimized code
I struggled to create a minimized version which doesn't use the library but I think it should be possible with the right combination of generics.

```scala
//> using scala 3.nightly
//> using dep cc.redberry:rings:2.5.8

import cc.redberry.rings.io.Coder

def main =
  new Coder[Int, Int, Int](null, null, null, null, null)
```
This is the class definition from the library https://github.com/PoslavskySV/rings/blob/develop/rings/src/main/java/cc/redberry/rings/io/Coder.java.

## Output

```scala
Error: error while loading Coder$1,
  class file cc/redberry/rings/io/Coder$1.class is broken (version 52.0),
  please check the JDK compatibility of your Scala version (3.10.0-RC1-bin-20260609-b34a019-NIGHTLY),
  reading aborted with class java.util.NoSuchElementException:
  key not found: Oth
[error] ./test1001.scala:7:7
[error] none of the overloaded alternatives named <init> can be accessed as a member of cc.redberry.rings.io.Coder[Int, Int, Int] from the top-level definitions in package <empty>.
[error]   private[io] constructor Coder can only be accessed from package cc.redberry.rings.io in package cc.redberry.rings.
[error]   private constructor Coder can only be accessed from class Coder in package cc.redberry.rings.io.
[error]   new Coder[Int, Int, Int](null, null, null, null, null)
[error]       ^^^^^^^^^^^^^^^^^^^^
```

## Expectation
Only 
```
[error] ./test1001.scala:7:7
[error] none of the overloaded alternatives named <init> can be accessed as a member of cc.redberry.rings.io.Coder[Int, Int, Int] from the top-level definitions in package <empty>.
[error]   private[io] constructor Coder can only be accessed from package cc.redberry.rings.io in package cc.redberry.rings.
[error]   private constructor Coder can only be accessed from class Coder in package cc.redberry.rings.io.
[error]   new Coder[Int, Int, Int](null, null, null, null, null)
[error]       ^^^^^^^^^^^^^^^^^^^^
```
Should not have the "classfile is broken" error.

## Comment
This library is used by SCAS which is in community build C. We run into it as soon as we need to calculate the constructor denotation for `Coder`. The SCAS library didn't previously require this for some reason, but we need it for #26156.

### PR DESCRIPTION
**Pull request by SolalPirelli**
Fixes #26312

## How much have you relied on LLM-based tools in this contribution?

Not at all

## How was the solution tested?

New automated tests (including the issue's reproducer, if applicable)

### CODE REVIEW

File: tests/neg/26312.check [hunk 1]
```diff
@@ -0,0 +1,10 @@
+-- [E173] Reference Error: tests/neg/26312/Use_2.scala:3:16 ------------------------------------------------------------
```
Comment (SolalPirelli) [inline tests/neg/26312.check:1]: without this fix, there's a nopos-error "broken classfile"

File: compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala [hunk 2]
```diff
@@ -579,9 +579,26 @@ class ClassfileParser(
...
+              reportErrors = false)
+        }
+        finally
+          index = savedIndex
+      }
+    }
     // Warning: sigToType contains nested completers which might be forced in a later run!
     // So local methods need their own ctx parameters.
-    def sig2type(tparams: immutable.Map[Name, Symbol], skiptvs: Boolean)(using Context): Type = {
+    def sig2type(tparams: immutable.Map[Name, Symbol], skiptvs: Boolean, isParent: Boolean = false)(using Context): Type = {
```
Comment (prolativ) [inline compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala]: I was just going to complain that the unfortunate thing about moving the definition of `var tparams` is that now `var tparams` is slightly more likely to be confused with `tparams` parameter of the methods `sig2type` and `sig2typeBounds`. Just after that I asked Claude for some help with the review and it spotted a different problem:

```
Consider this Java:

////////

public class Wrapper<A, B> { }

class Outer {
    public <Oth> Wrapper<Oth, Oth> make() {
        return new Wrapper<Oth, Oth>() { };  // anonymous class Outer$1
    }
}

////////

Outer$1's classfile signature would be something like:

LWrapper<TOth;TOth;>;

When the parser hits the 'L' case for Wrapper and enters processTypeArgs, it loops over the two type arguments calling sig2type(tparams, skiptvs, isParent) for each:

// First iteration: index at TOth;
sig2type(...) → tparams.get("Oth") == None → creates symbol S1, adds to classTParams

// Second iteration: index at TOth; again
sig2type(...) → tparams.get("Oth") == None  ← still None, tparams unchanged
             → creates symbol S2, enters S2 into owner.asClass again

Both S1 and S2 get entered into the owner class's scope under the same name Oth. Whether this actually causes a problem depends on how owner.asClass.enter handles duplicate names — it may silently overwrite or accumulate both, potentially confusing later type-checking of that class.
```

Unless I missed something, getting rid of `tparams` parameter in favour of referencing the `tparams` variable directly combined with updating `tparams` as well when `classTParams` is updated with a new symbol below should solve both problems a the same time.

Also, probably it would be worth to add a test for handling such a duplicate type param reference

Comment (SolalPirelli) [reply compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala]: Thanks, I forgot about that.

I tried refactoring the tparams stuff but it's quite confusing and a simple attempt causes a couple tests to fail with clearly generic type parameters related stuff. What exactly the two `tparams` are supposed to mean isn't obvious.

So I added an `orElse(classTParams.get(n))` here instead, which fixes the problem without broader repercussions.

Comment (prolativ) [reply compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala]: Out of curiosuty: which tests failed after the refactor I suggested? This seems quite strange given that the actual argument given for the value of `tparams` parameter seems to always be the value of `tparams` variable without any modifications

Comment (SolalPirelli) [reply compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala]: Hmm, I think I tried to do too much last time, or maybe I was just tired on a Friday afternoon ^^' seems to work now, thanks.

**Review by prolativ — APPROVED** [review]
